test/e2e/info_api_test.go

```diff
@@ -0,0 +1,194 @@
+	testutil.Ok(t, e2e.StartAndWaitReady(prom1, sidecar1, prom2, sidecar2, prom3, sidecar3))
+
+	const bucket = "info-api-test"
+	m := e2ethanos.NewMinio(e, "thanos-minio", bucket)
+	testutil.Ok(t, e2e.StartAndWaitReady(m))
+	store, err := e2ethanos.NewStoreGW(
+		e,
+		"1",
+		client.BucketConfig{
+			Type: client.S3,
+			Config: s3.Config{
+				Bucket:    bucket,
+				AccessKey: e2edb.MinioAccessKey,
+				SecretKey: e2edb.MinioSecretKey,
+				Endpoint:  m.InternalEndpoint("http"),
+				Insecure:  true,
+			},
+		},
+		"",
+	)
+	testutil.Ok(t, err)
+	testutil.Ok(t, e2e.StartAndWaitReady(store))
+
+	// Register `sidecar1` in all flags (i.e. '--store', '--rule', '--target', '--metadata', '--exemplar', '--endpoint') to verify
+	// '--endpoint' flag works properly works together with other flags ('--target', '--metadata' etc.).
+	// Register 2 sidecars and 1 storeGW using '--endpoint'.
+	// Register `sidecar3` twice to verify it is deduplicated.
+	q, err := e2ethanos.NewQuerierBuilder(e, "1", sidecar1.InternalEndpoint("grpc")).
+		WithTargetAddresses(sidecar1.InternalEndpoint("grpc")).
+		WithMetadataAddresses(sidecar1.InternalEndpoint("grpc")).
+		WithExemplarAddresses(sidecar1.InternalEndpoint("grpc")).
+		WithRuleAddresses(sidecar1.InternalEndpoint("grpc")).
+		WithEndpoints(
+			sidecar1.InternalEndpoint("grpc"),
+			sidecar2.InternalEndpoint("grpc"),
+			sidecar3.InternalEndpoint("grpc"),
+			sidecar3.InternalEndpoint("grpc"),
+			store.InternalEndpoint("grpc"),
+		).
+		Build()
+	testutil.Ok(t, err)
+	testutil.Ok(t, e2e.StartAndWaitReady(q))
+
+	expected := map[string][]query.EndpointStatus{
+		"sidecar": {
+			{
+				Name: "e2e_test_info-sidecar-alone1:9091",
+				LabelSets: []labels.Labels{{
+					{
+						Name:  "prometheus",
+						Value: "prom-alone1",
+					},
+					{
+						Name:  "replica",
+						Value: "0",
+					},
+				}},
+			},
+			{
+				Name: "e2e_test_info-sidecar-alone2:9091",
+				LabelSets: []labels.Labels{{
+					{
+						Name:  "prometheus",
+						Value: "prom-alone2",
+					},
+					{
+						Name:  "replica",
+						Value: "0",
+					},
+				}},
+			},
+			{
+				Name: "e2e_test_info-sidecar-alone3:9091",
+				LabelSets: []labels.Labels{{
+					{
+						Name:  "prometheus",
+						Value: "prom-alone3",
+					},
+					{
+						Name:  "replica",
+						Value: "0",
+					},
+				}},
+			},
+		},
+		"store": {
+			{
+				Name:      "e2e_test_info-store-gw-1:9091",
+				LabelSets: []labels.Labels{},
+			},
+		},
+	}
+
+	url := "http://" + path.Join(q.Endpoint("http"), "/api/v1/stores")
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancel()
+
+	err = runutil.Retry(time.Second, ctx.Done(), func() error {
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+		if err != nil {
+			return err
+		}
+
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			return err
+		}
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+
```

Comment:
We should also do `defer runutil.CloseWithErrCapture(&err, resp.Body, "response body close")`